### PR TITLE
Make sure to save Post when dismissing the EditPostActivity

### LINKED ISSUE #8570: Drafts Automatically & Unintentionally Publishing
I'm not too sure if this is the best place to report an issue, but there's something which seems to have happened more than once, and it's rather frustrating.

### Expected behaviour

Drafts should remain drafts when media is uploaded to the Media Library, but the media should be uploaded regardless

### Actual behaviour

A recent draft is published when media is uploaded to the Media Library, even though there was no modification made to that post. 

### Steps to reproduce the behaviour

1. Create a new blog post and upload an image
2. Close the app (do not save the blog post as a draft, but do not click "Publish" either)
3. You can wait if you wish (there was a matter of days between step 2 and step 4 in the times which this has happened to me)
4. Enter your Gallery and hold an item, then click "Share" or the equivalent
5. Select "WordPress" and "Add to Media Library"
6. Draft turns into a published post

### Additional Context

-Samsung Galaxy Tab S2 SM-T713
-Version 11.0 of WordPress app 
-Using the Beta editor
-Activity log, unfortunately I can't upload directly to GitHub for some reason :(
https://clubpenguinmountains.files.wordpress.com/2018/11/screenshot_20181107-182850.jpg

An issue similar to this occurred in February 2018 too with my blog, and if I recall correctly, it was this exact same thing too which provoked the post to publish unintentionally. I'm assuming that because of this, it is a bug with the WordPress app, as it's never happened through a web browser, including with Calyspo and the Classic Editor. I'm able to recreate this on a test site too.

Thanks! :)

### PR DESCRIPTION
**Pull request by mzorz**
Fixes #8570 

This PR makes the save logic run in `onDestroy()` to make sure the `UploadService` gets the latest copy of the Post if running in parallel as the `EditPostActivity` is killed.

To test:
1. start a draft
2. add an image (preferrably something large so it takes time to upload)
3. while the image is uploading, tap the recent apps list button on Android and dismiss the app
4. observe the upload notification continues to upload
5. open the app again, observe the Post is a draft.

Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

### CODE REVIEW

**Comment by mzorz** [comment]
Adding analysis information here for future reference.

Steps to reproduce the issue / diagnosis by @khaykov (before the fix):

> Start a new post, add media to it, force-close app (through recent items button, for example), open the app again and you will notice that the post from editor got published.
> Publishing happens when `onDestroy` of `UploadService` is called.
> 

The primary problem here is that we're using the `PUBLISHED` state as per default when instantiating a new Post, but this has historical reasons and there's a lot of logic heavily relying on this assumption in `EditPostActivity`. There are plans to refactor this and most probably will be taken care of once Gutenberg is the only available editor, as there's much logic in there that makes checks for Visual, Legacy, and Classic editors (plus now Gutenberg).
So, given changing this assumption will probably break something else, for now I've decided to stick with it and produce a fix for the condition we're running into between the `Editor` and the `UploadService`, which I'll describe following.

Ok so: what happens here is that when the editor is open we mark this current post as opened and prevent the UploadService from doing anything with that unless the user requests (i.e. publishing or updating). The current opened post flag is cleared when we exit the editor, so we can let the `UploadService` do its thing. This condition shouldn't happen as we're bailing from the final processing of posts in `UploadService` when a Post is being currently open in the Editor, but there's a race condition that happens when you delete the app from the recents list that makes the `EditPostActivity` (Editor) be killed first, its `onDestroy()` callback gets run (and that's where we remove the sticky flag in `removePostOpenInEditorStickyEvent()`), and then only after that the `UploadService`'s `onDestroy()` gets called (as it finished uploading media), which makes the final processing, checks that the Post is not currently being edited (as the flag has been removed), and finds the Post to be all good for upload. It also finds the Post because it was registered for having media attached (Posts with no media wouldn't run into this condition).
Then, given the `default` status for a Post historically has been `PUBLISHED`, it gets uploaded _and_ published on the web.

Also, the `UploadService` may be destroyed way after the main app process.

My first attempt to fix was to do the following:
- removing the `PostOpenedInEditor` sticky event not in `onDestroy`, but added it right before every `finish()` call to remove it when the user actually won't be editing anymore (pressed back, finished publishing voluntarily, or the Editor had to bail because some of some bad parameter on creation) but not because of an external destroying cause.


However, this does not prevent the **_!!previous!!_** Post from being uploaded on next time you trigger the `UploadService`. For example:
1. start a draft
2. add an image
3. dismiss the app
4. at this point the upload service continues to upload media and once it's finished, it checks for media/post registered posts, finds your post and checks the opened post sticky event flag
5. given the flag is marked as open, it doesn't push the Post, so we're good. But, the post will still be registered in the media/post FluxC registry.

Here comes the good part:

6. now, re-open the app, and start a **new** draft. This will make this new draft update the opened post sticky event flag
7. again, add an image, this will register the Post with FluxC for async media upload
8. dismiss the app
9. the media will continue to upload, once it's finished, the same check as in  4 is made: the **_previous_** post (started in step 1 above) is found from the FluxC registry, we check if it's open _and  of course **it's not**_, so we end up uploading it!, which is not what we want.

**In summary**: any previous posts that are registered with media in FluxC, with exception of the latest Post you're editing, will be uploaded by the UploadService. And for any posts that have not exited the editor gracefully, these will have been marked `PUBLISHED` and uploaded as such.

So, unfortunately we couldn't use this approach, as at the point the `doFinalProcessingOfPosts` happens in the UploadService, we can't know whether the user wanted to PUBLISH or it was just the default state of the Post.

The fix for this is then, to make sure to run the status change logic in `onDestroy()`, to update the Post before the `UploadService` gets to it. 

Thank you @khaykov for diving into this and providing with pointers / context to the problem!

**Comment by khaykov** [comment]
@mzorz Thank you so much for a quick response on this one! I actually tried this fix before, and while it does indeed work in a scenario you described, the draft will still get published if you wait for the media to finish uploading before the force-closing the app.

**Comment by mzorz** [comment]
> the draft will still get published if you wait for the media to finish uploading before the force-closing the app.

oh you're right! Let me try fix that

**Comment by mzorz** [comment]
Closing in favor of https://github.com/wordpress-mobile/WordPress-Android/pull/9322
